go/consensus/tendermint/abci/context_test.go

```diff
@@ -0,0 +1,82 @@
+package abci
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestContext(t *testing.T) {
+	require := require.New(t)
+
+	now := time.Unix(1580461674, 0)
+	appState := NewMockApplicationState(MockApplicationStateConfig{})
+	ctx := appState.NewContext(ContextBeginBlock, now)
+	defer ctx.Close()
+
+	// Add some state.
+	tree := ctx.State()
+	err := tree.Insert(ctx, []byte("key"), []byte("value"))
+	require.NoError(err, "Insert")
+
+	// Test checkpoints.
+	cp := ctx.StartCheckpoint()
+	// Should panic on nested checkpoints.
+	require.Panics(func() { ctx.StartCheckpoint() })
+	overlay := ctx.State()
+	require.NotEqual(&tree, &overlay, "new State() should return the overlay")
+
+	// Existing state should be there.
+	value, err := overlay.Get(ctx, []byte("key"))
+	require.NoError(err, "Get")
+	require.EqualValues([]byte("value"), value)
+
+	// Add some state to the overlay.
+	err = overlay.Insert(ctx, []byte("blah"), []byte("value2"))
+	require.NoError(err, "Insert")
+	err = overlay.Remove(ctx, []byte("key"))
+	require.NoError(err, "Remove")
+
+	// Make sure updates didn't leak.
+	value, err = tree.Get(ctx, []byte("key"))
```

Comment:
> is this a new behavior, that you can put a "State" result in your pocket and have access to a snapshot from the checkpoint?

So while you can retain a reference to the previous state, but if you modify it, these updates may leak to the checkpoint, so you shouldn't do that. We could try to prevent it by somehow invalidating all previous references but I'm not sure if it is worth the trouble.

> And you could update that outer tree while in the middle of a transaction.

As mentioned above, any updates to the outer tree may leak to the pending tree.

Reply:
I've added an explicit comment to `StartCheckpoint` mentioning previous references to `State`.